clustering_code/edchan-kmeans.py

```diff
@@ -0,0 +1,913 @@
+    CPL_VSIL_CURL_ALLOWED_EXTENSIONS='TIF'
+    GDAL_HTTP_UNSAFESSL='YES'
+"""
+load_dotenv() # Load .env file
+
+data_dir = os.path.join(et.io.HOME, et.io.DATA_NAME)
+cache_dir = os.path.join(data_dir, 'clustering-kmeans')
+
+if not os.path.exists(cache_dir):
+    os.makedirs(cache_dir)
+
+# %%
+username = os.getenv('EARTHDATA_USERNAME')
+password = os.getenv('EARTHDATA_PASSWORD')
+
+# earthaccess.login doesn't seem to be creating this file
+netrc_fpath = os.path.expanduser("~/.netrc")
+if not netrc_fpath:
+    with open(netrc_fpath, 'w') as file:
+        file.write(f"machine urs.earthdata.nasa.gov\nlogin {username}\npassword {password}\n")
+
+# %% [markdown]
+# Below you can find code for a caching **decorator** which you can use in your code. To use the decorator:
+# 
+# ```python
+# @cached(key, override)
+# def do_something(*args, **kwargs):
+#     ...
+#     return item_to_cache
+# ```
+# 
+# This decorator will **pickle** the results of running the `do_something()` function, and only run the code if the results don't already exist. To override the caching, for example temporarily after making changes to your code, set `override=True`.
+
+# %%
+def cached(key, override=False):
+    """"""
+    def compute_and_cache_decorator(compute_function):
+        """"""
+        def compute_and_cache(*args, **kwargs):
+            """Perform a computation and cache, or load cached result"""
+            filename = os.path.join(et.io.HOME, et.io.DATA_NAME, 'jars', f'{key}.pickle')
+            
+            # Check if the cache exists already or override caching
+            if not os.path.exists(filename) or override:
+                # Make jars directory if needed
+                os.makedirs(os.path.dirname(filename), exist_ok=True)
+                
+                # Run the compute function as the user did
+                result = compute_function(*args, **kwargs)
+                
+                # Pickle the object
+                with open(filename, 'wb') as file:
+                    pickle.dump(result, file)
+            else:
+                # Unpickle the object
+                with open(filename, 'rb') as file:
+                    result = pickle.load(file)
+                    
+            return result
+        
+        return compute_and_cache
+    
+    return compute_and_cache_decorator
+
+# %% [markdown]
+# ## STEP 2: STUDY SITE
+
+# %% [markdown]
+# For this analysis, you will use a watershed from the [Water Boundary Dataset](https://www.usgs.gov/national-hydrography/access-national-hydrography-products), HU12 watersheds (WBDHU12.shp).
+# 
+# YOUR TASK:
+#   1. Download the Water Boundary Dataset for region 8 (Mississippi)
+#   2. Select watershed 080902030506
+#   3. Generate a site map of the watershed
+
+# %%
+def find_file(pattern, path):
+    return glob.glob(f'{path}/**/{pattern}', recursive=True)
+def copy_file(src, dst):
+    shutil.copy(src, dst)
+
+
+wbdhu12_shp = os.path.join(cache_dir, 'WBDHU12.shp')
```

Comment:
I don't think you necessarily need more comments here to explain how the code works since it's pretty clear to my reading, but it can be nice in these scientific workflow to give some clues about where you found the data.